File: desktop/src/lib/seat-selftest.js
const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");
const { PACK_IDS, renderPack, activationReply } = require("./seat-packs");
const runtime = require("./seat-runtime");

function assert(cond, message) {
  if (!cond) throw new Error(message);
}

function main() {
  const stamps = {
    codex: "ASTRA//UNLOCK",
    claude: "ANTHRO//CLEAR",
    grok: "GROK//OPEN",
    deepseek: "SEEK//BARE",
    glm53: "GLM//STRAIGHT",
    gemini: "Gemini 席位上的冷咖啡执行核",
  };
  const rendered = {};
  for (const id of PACK_IDS) {
    rendered[id] = renderPack(id);
    assert(rendered[id].includes(stamps[id]), `${id} pack missing original stamp`);
    assert(rendered[id].includes("<!-- CHA-"), `${id} missing marker`);
  }
  const unique = new Set(Object.values(rendered));
  assert(unique.size === PACK_IDS.length, "packs must be unique");
  const page = activationReply();
  assert(page.includes("QQ一群 1057540028"), "activation page missing QQ 1");
  assert(page.includes("QQ二群 1077074552"), "activation page missing QQ 2");
  assert(page.includes("来杯冰美式，然后开始你的开心人生吧。"), "activation page missing ice americano line");
  for (const id of PACK_IDS) {
    assert(rendered[id].includes("1057540028"), `${id} missing QQ 1`);
    assert(rendered[id].includes("1077074552"), `${id} missing QQ 2`);
    assert(rendered[id].includes("开心人生"), `${id} missing manifesto`);
  }

  const root = fs.mkdtempSync(path.join(os.tmpdir(), "cha-seats-"));
  try {
    for (const id of PACK_IDS) {
      const home = path.join(root, id);
      const pre = runtime.preview(id, home);
      assert(pre.ok && pre.text, `${id} preview failed`);
      const dep = runtime.deploy(id, home);
      assert(dep.ok && dep.writes.length, `${id} deploy failed`);
      const ver = runtime.verify(id, home);
      assert(ver.ok, `${id} verify failed: ${JSON.stringify(ver.checks)}`);
      const res = runtime.restore(id, home);
      assert(res.ok, `${id} restore failed`);
    }
  } finally {
    fs.rmSync(root, { recursive: true, force: true });
  }
  process.stdout.write(`seat-selftest ok · ${PACK_IDS.join(",")}\n`);
}

if (require.main === module) main();
module.exports = { main };

File: tools/seat_selftest.py
#!/usr/bin/env python3
"""Temp-home roundtrip for all six CHA seats."""

from __future__ import annotations

import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from seats import PACK_IDS, deploy, preview, render_pack, restore, verify  # noqa: E402

STAMPS = {
    "codex": "ASTRA//UNLOCK",
    "claude": "ANTHRO//CLEAR",
    "grok": "GROK//OPEN",
    "deepseek": "SEEK//BARE",
    "glm53": "GLM//STRAIGHT",
    "gemini": "Gemini 席位上的冷咖啡执行核",
}


def main() -> int:
    bodies = {seat: render_pack(seat) for seat in PACK_IDS}
    if len(set(bodies.values())) != len(PACK_IDS):
        raise SystemExit("packs are not unique")
    for seat, stamp in STAMPS.items():
        if stamp not in bodies[seat]:
            raise SystemExit(f"{seat} missing original stamp")
    with tempfile.TemporaryDirectory(prefix="cha-seats-") as raw:
        root = Path(raw)
        for seat in PACK_IDS:
            home = root / seat
            pre = preview(seat, home)
            if not pre.get("ok"):
                raise SystemExit(f"{seat} preview failed")
            dep = deploy(seat, home)
            if not dep.get("ok"):
                raise SystemExit(f"{seat} deploy failed")
            ver = verify(seat, home)
            if not ver.get("ok"):
                raise SystemExit(f"{seat} verify failed: {ver}")
            res = restore(seat, home)
            if not res.get("ok"):
                raise SystemExit(f"{seat} restore failed")
    print("seat_selftest ok · " + ",".join(PACK_IDS))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
